### PR TITLE
[PLDM FRU]: Removing FRU Record from FRU Record Table

### PR DESCRIPTION
**Pull request by ArchanaKakani**
When a CEC FRU such as a Fan or Power Supply is concurrently removed,
BMC properly sends event to remove the FRU Record Set ID and to
 modify the Entity Association PDRs.
BMC also removes the corresponding state sensors/ effecters PDRs.
However, BMC does not remove the FRU entry from the FRU Record Table

Modified code to delete FRU Record from FRU Record table.

Tested: Using busctl command and pldmtool
 busctl set-property xyz.openbmc_project.Inventory.Manager
      /xyz/openbmc_project/inventory/system/chassis/motherboard/fan0
     xyz.openbmc_project.Inventory.Item Present b false

 pldmtool fru GetFruRecordTable -v

Signed-off-by: ArchanaKakani <archana.kakani@ibm.com>

### CODE REVIEW

**Review by sagisin — APPROVED** [review]
looks good to me

File: libpldmresponder/fru.cpp [hunk 1]
```diff
@@ -418,6 +418,7 @@ void FruImpl::removeIndividualFRU(const std::string& fruObjPath)
               << removeEntity.entity_container_id << " ]\n";
     associatedEntityMap.erase(fruObjPath); // sm00
 
+    deleteFruRecord(rsi);
     /* if (table
```
Comment (jaypadath) [inline libpldmresponder/fru.cpp:422]: Does this commented explanation makes any sense as we are deleting the fru record table entry now ?

File: libpldmresponder/fru.cpp [hunk 2]
```diff
@@ -453,6 +454,43 @@ void FruImpl::removeIndividualFRU(const std::string& fruObjPath)
...
+    const struct pldm_fru_record_tlv* tlv;
+    size_t pos = 0;
+
+    while (table.size() > pos)
+    {
+        size_t recordLen = sizeof(struct pldm_fru_record_data_format) -
+                           sizeof(struct pldm_fru_record_tlv);
+        tlv = recordSetSrc->tlvs;
+
+        for (int i = 0; i < recordSetSrc->num_fru_fields; i++)
```
Comment (jaypadath) [inline libpldmresponder/fru.cpp:473]: Please use unsingned int (uint) instead of signed one

Comment (ArchanaKakani) [reply libpldmresponder/fru.cpp]: Done

File: libpldmresponder/fru.cpp (hunk 1 above)
Comment (ArchanaKakani) [reply libpldmresponder/fru.cpp]: Removed the comment

File: libpldmresponder/fru.cpp [hunk 3]
```diff
@@ -399,8 +399,6 @@ void FruImpl::removeIndividualFRU(const std::string& fruObjPath)
      free(out);*/
     // sm00
 
-    pldm_entity_association_tree_delete_node(entityTree, removeEntity);
```
Comment (Pavithrab7) [inline libpldmresponder/fru.cpp:402]: why is this removed?

File: libpldmresponder/fru.cpp [hunk 4]
```diff
@@ -453,6 +436,43 @@ void FruImpl::removeIndividualFRU(const std::string& fruObjPath)
...
+            size_t len = sizeof(*tlv) - 1 + tlv->length;
+            recordLen += len;
+            tlv = reinterpret_cast<const struct pldm_fru_record_tlv*>(
+                (char*)tlv + len);
+        }
+        if ((recordSetSrc->record_set_id != htole16(rsi) && rsi != 0))
+        {
+            std::copy(table.begin() + pos, table.begin() + pos + recordLen,
+                      std::back_inserter(updatedFruTbl));
+        }
```
Comment (manojkiraneda) [inline libpldmresponder/fru.cpp:466]: we are copying everything from begin to the postiion , but what about the records that are after the position ?
if the fru to be removed is at 5th postion , we just have to skip 5 an copy the entire table, but looking at this i feel we are copying only till 5th location.

Am i missing something here ?

**Review by manojkiraneda — CHANGES_REQUESTED** [review]

File: libpldmresponder/fru.cpp (hunk 4 above)
Comment (ArchanaKakani) [reply libpldmresponder/fru.cpp:466]: We r running this if block in the while loop till we reach end of the frutable. So it will copy all other FRU records except the rsi matching ones.

File: libpldmresponder/fru.cpp [hunk 5]
```diff
@@ -453,6 +436,43 @@ void FruImpl::removeIndividualFRU(const std::string& fruObjPath)
...
+            table.data());
+
+    const struct pldm_fru_record_tlv* tlv;
+    size_t pos = 0;
+
+    while (table.size() > pos)
+    {
+        size_t recordLen = sizeof(struct pldm_fru_record_data_format) -
+                           sizeof(struct pldm_fru_record_tlv);
+        tlv = recordSetSrc->tlvs;
```
Comment (manojkiraneda) [inline libpldmresponder/fru.cpp:453]: null check missing before de-reference.

**Review by manojkiraneda — APPROVED** [review]

File: libpldmresponder/fru.cpp (hunk 3 above)
Comment (ArchanaKakani) [reply libpldmresponder/fru.cpp:397]: Done
